accepted/2021/workloads/workload-installation.md

```diff
@@ -0,0 +1,338 @@
+
+In order to support these different installers, the .NET SDK will have an "installer abstraction layer" (IAL) which will allow common installation code to work with the different types of installers.  (NOTE: We've previously called this a Platform Abstraction Layer (PAL), so some documents may continue to refer to it as such.)
+
+The IAL is an abstraction that is internal to the .NET SDK.  It is not intended as a public extensibility mechanism to allow supporting arbitrary installers, and we may make breaking changes to its API as needed.
+
+# Supported installation types
+
+We plan to (eventually) support the following installation types.  Some of them may not be supported in .NET 6 if we don't have workloads in .NET 6 that are supported on the corresponding platforms.
+
+## .NET SDK managed installation
+
+In this implementation of the IAL, the .NET SDK itself manages workload installations directly by writing files to the dotnet folder.  This is generally for copies of the .NET SDK which were not installed via an OS installer, for example those that are installed via the dotnet-install scripts, or via unzipping archives of the .NET SDK.
+
+Other terms we have at times used for this installation type: xcopy, universal installer, files on disk, NuGet installer
+
+## Visual Studio managed installation
+
+This is an installation of the .NET SDK installed and managed by the Visual Studio installer.  The .NET SDK workloads will have corresponding Visual Studio workloads or components, which will install workload packs via Windows MSIs.  Note that the Visual Studio workloads or components should also include the necessary Visual Studio components (designers, project system, etc) to support the workload, in addition to the .NET SDK workload packs.
+
+## Standalone Windows installation
+
+This is an installation of the .NET SDK installed on Windows via the .NET SDK installer bundle, which is the .exe which can be downloaded on the .NET SDK download pages.  The bundle installs the same MSIs for the .NET SDK as Visual Studio does, and the workload installer will likewise install the same MSIs for workload packs as the Visual Studio installer.
+
+The same dotnet folder may have different versions of the SDK installed by both the Visual Studio installer and the Standalone installer, which may share the same versions of some components.  For the .NET SDK, the underlying MSIs are ref-counted so that they are uninstalled when there are no more installations that depend on them.  We will need to make sure that workload installation and management also works well when there are different install types coexisting.
+
+## Package manager installation
+
+This is a copy of the .NET SDK installed as a package from a package manager such as APT or RPM.  In this case, the workloads and / or workload packs will also be represented as packages which will be installed via the package manager.
```

Comment:
Non source-buildable workloads should be usable with a source-built .NET installation.

Preferably this doesn't require the user to add Microsoft specific feeds, but such packages can be fetched directly from nuget.org.

Such workloads shouldn't install to the source-build DOTNET_ROOT. This is meant for system packages. Maybe they can install to a user-specific location?
That would also make them installable by non-root.

Reply:
> That would also make them installable by non-root.

Some containers don't run as root, so that may be another reason to support installing workloads by non-root.

Reply:
That can only work for packages we can source-build.

The location .NET is installed is meant for package installs only.
This location is different between Microsoft and Red Hat packages. It's even different for Fedora and RHEL packages.

Reply:
@dsplaisted because not all workloads will be source-buildable, I think this means the workloads need to be installable under `$HOME`.

Reply:
@tmds Thanks for the feedback, it sounds like we will need to support installing workloads outside of the DOTNET_ROOT.  However, we don't currently plan to do this in .NET 6.

That would mean that for Blazor AoT in .NET 6, developers using source-built Linux would either need to run the workload install elevated and the workload would be copied into DOTNET_ROOT, or Blazor AoT wouldn't be supported at all.

Does that sound like something we can live with for .NET 6?

In .NET 7 we expect to move more functionality to workloads, and would need to support installing them outside of DOTNET_ROOT.

Reply:
With https://github.com/dotnet/sdk/pull/14841 there's a way to override the workload lookup paths so theoretically something could be hacked to make it work.

Reply:
> using source-built Linux would either need to run the workload install elevated and the workload would be copied into DOTNET_ROOT

This doesn't work because the .NET builds running in containers, for security, are performed as non-root.

Also, manually installing into these locations is an anti-pattern.

And, for any workload, the user experience of an open-source built .NET would be broken compared to Microsoft's .NET build.

Reply:
@tmds I'm trying to figure out what's OK for .NET 6, where we aren't planning to support installing workloads outside of DOTNET_ROOT.

@akoeplinger reminded me of another option, which is to use environment variables that were added as test hooks in order to load workloads from arbitrary folders.  This could be an option for source-built .NET.  However, workload installation (`dotnet workload install`) wouldn't support this.  So you would need to download the right NuGet packages, extract them into the right folders, and set the right environment variables to point to those folders.  You might be able to automate this with a script.

Does that sound like an acceptable solution for .NET 6?

Reply:
That is a broken user experience compared to Microsoft's build.

Microsoft aim for source-built to be as capable as Microsoft's build.

Is it feasible to make this work for .NET 6?

Reply:
@tmds @omajid I've added more detail to the package manager section about what we expect to do for .NET 6 and .NET 7.

For .NET 6, I don't think we have the capacity to support `dotnet install` installing workloads outside of the .NET SDK folder.

Reply:
@dsplaisted we need to do better.

It's not ok for Microsoft proprietary builds to deliver more features than open-source built .NET.

Reply:
@dsplaisted Microsoft tells us they want source-build to be as capable as their build, and they want it to become the default mechanism for delivering .NET on Linux.

At the same time, we run into situations were things are not designed or implemented to work well with source-build.

We should really avoid to getting into this situation and it should not be considered acceptable by anyone that source-build users get a degraded experience.